### PR TITLE
[linstor] Change csi endpoint to replicated.csi.deckhouse.io

### PR DESCRIPTION
**Pull request by AleksZimin**
## Description

This PR updates the CSI driver by changing its endpoint from `linstor.csi.linbit.com` to `replicated.csi.storage.deckhouse.io` and introduces an automated migration strategy for a seamless transition.

It's important to note that the migration includes restarting the **kubelet** on all nodes running `linstor-csi-node` pods and will temporarily switch off the LINSTOR control plane. This process is designed to ensure a successful update with minimal operational impact. Users should prepare for this brief period of adjustment.

**Attention!** Reverse migration is not supported. After completing the migration, attempting to switch back to the LINSTOR module will result in the LINSTOR control plane becoming non-operational.

## Why do we need it, and what problem does it solve?

Switching to a new endpoint is a crucial step in our move away from the LINSTOR backend towards deploying our own storage controller for Kubernetes, based on DRBD and LVM. This change is made to  simplify the migration process to our controller in future and to ensure that the future replacement of the LINSTOR CSI driver with our own will be seamless for users.

## What is the expected result?

The successful completion of the migration will be marked by the CSI driver's endpoint update and the creation of a secret, named csi-migration-finished, in the d8-sds-drbd namespace. This secret serves as an indicator of successful migration. To verify its existence, you can use the following command:
```bash
kubectl get secret csi-migration-finished -n d8-sds-drbd
```

The migration script will update the endpoint in StorageClasses, PVs, PVCs, VolumeAttachments, VolumeSnapshotClasses and VolumeSnapshotContents ensuring all relevant resources reflect the new endpoint.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: hooks/migrate_csi_endpoint.sh [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+  if kubectl get secret "$SECRET_NAME" -n "$NAMESPACE" > /dev/null 2>&1; then
+    echo "Secret ${NAMESPACE}/${SECRET_NAME} exists. Migration has been finished before"
+    values::set sdsReplicatedVolume.internal.csiMigrationHook.completed "true"
+    exit 0
+  fi
+
+  echo "Secret ${NAMESPACE}/${SECRET_NAME} does not exist. Starting csi migration"
+
+  sc_list=$(kubectl get sc -o json | jq -r ".items[] | select(.provisioner == \"$old_driver_name\") | .metadata.name")
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:45]:
```suggestion
  sc_list=$(kubectl get sc -o json | jq -r --arg oldDriverName "$old_driver_name" `.items[] | select(.provisioner == $oldDriverName) | .metadata.name`)
```

File: hooks/migrate_csi_endpoint.sh [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+# limitations under the License.
+
+source /deckhouse/shell_lib.sh
+
+get_config() {
+  echo '{"configVersion":"v1", "beforeHelm": 10}'
+}
+
+run_trigger() {
+  export old_driver_name="linstor.csi.linbit.com"
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:24]: Vars should be CAPSed imho.

File: hooks/migrate_csi_endpoint.sh [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+  export new_driver_name="replicated.csi.storage.deckhouse.io"
+  export old_attacher="linstor-csi-linbit-com"
+  export new_attacher="replicated-csi-storage-deckhouse-io"
+  export LABEL_KEY="storage.deckhouse.io/need-kubelet-restart"
+  export LABEL_VALUE=""
+  export SECRET_NAME="csi-migration-finished"
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:30]:
```suggestion
  export SECRET_NAME="csi-migration-completed"
```
Completed is better, and there is value `internal.csiMigrationHook.completed`.

File: hooks/migrate_csi_endpoint.sh [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    exit 0
+  fi
+
+  echo "Secret ${NAMESPACE}/${SECRET_NAME} does not exist. Starting csi migration"
+
+  sc_list=$(kubectl get sc -o json | jq -r ".items[] | select(.provisioner == \"$old_driver_name\") | .metadata.name")
+  pvc_pv_list=$(kubectl get pv -o json | jq -r --arg oldDriverName "$old_driver_name" '.items[] | select(.spec.csi.driver == $oldDriverName) | .spec.claimRef.namespace + "/" + .spec.claimRef.name + "/" + .metadata.name')
+  pvc_list=$(kubectl get pvc --all-namespaces -o json | jq -r --arg oldDriverName "$old_driver_name" '.items[] | select(.metadata.annotations["volume.kubernetes.io/storage-provisioner"] == $oldDriverName) | .metadata.namespace + "/" + .metadata.name')
+  
+  if [[ -z "$sc_list" && -z "$pvc_pv_list" && -z "$pvc_list" ]]; then
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:49]: I would prefer counting the sc/pv/pvc items and put the counts in vars pv_pvc_count/pv_count/sc_count.

File: hooks/migrate_csi_endpoint.sh [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+  export temp_dir=$(mktemp -d)
+  cd "$temp_dir"
+  echo "temporary dir: $temp_dir"
+
+  migrate_storage_classes
+  migrate_pvc_pv
+  migrate_volume_snapshot_classes
+  migrate_volume_snapshot_contents
+  
+  nodes_with_volumes=$(kubectl get volumeattachments -o=jsonpath='{range .items[*]}{@.metadata.name}{"\t"}{@.spec.attacher}{"\t"}{@.status.attached}{"\t"}{@.spec.nodeName}{"\n"}{end}' | grep "${old_driver_name}" | grep 'true' | awk '{print $4}' | sort | uniq)
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:73]:
```suggestion
  nodes_with_volumes=$(kubectl get volumeattachments -o=json | jq -r --arg oldDriverName "${old_driver_name}" '.items[] | select(.spec.attacher == $oldDriverName) | .spec.nodeName' | sort | uniq)
  
```

File: hooks/migrate_csi_endpoint.sh [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+  delete_old_volume_attachments
+
+
+  for node in $nodes_with_volumes; do
+    echo "Add label ${LABEL_KEY}=${LABEL_VALUE} to node $node"
+    kubectl label node $node ${LABEL_KEY}=${LABEL_VALUE} --overwrite
+  done
+
+  echo affected_pvs_hash="$affected_pvs_hash"
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:84]: isn't obvious where was var set.

File: hooks/migrate_csi_endpoint.sh [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    return
+  fi
+
+  echo "Scaling down pods with label app=$APP_NAME in namespace $NAMESPACE"
+  kubectl scale deployment -n "$NAMESPACE" --replicas=0 "$APP_NAME"
+  wait_for_pods_scale_down "$NAMESPACE" "$APP_NAME"
+}
+
+migrate_storage_classes() {
+  sc_list=$(kubectl get sc -o json | jq -r ".items[] | select(.provisioner == \"$old_driver_name\") | .metadata.name")
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:122]: --arg oldDriverName

File: hooks/migrate_csi_endpoint.sh (hunk 4 above)
Comment (apolovov) [reply hooks/migrate_csi_endpoint.sh:49]: Got the idea, nevermind.

File: hooks/migrate_csi_endpoint.sh [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+      old_ifs=$IFS
+      IFS='/' read -r -a array <<< "$pvc_pv"
+      namespace=${array[0]}
+      pvc=${array[1]}
+      pv=${array[2]}
+
+      kubectl get pvc $pvc -n $namespace -o yaml > pvc-${pvc}.yaml
+      kubectl get pv $pv -o yaml > pv-${pv}.yaml
+
+      cp pvc-${pvc}.yaml old.pvc-${pvc}.yaml.$timestamp
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:172]: are you sure you don't need to sed the pvc...yaml file?

File: hooks/migrate_csi_endpoint.sh [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+      echo "Deleting finalizer from pv: $pv"
+      kubectl patch pv $pv --type json -p '[{"op": "remove", "path": "/metadata/finalizers"}]'
+      echo "Patching annotations in pvc: $pvc"
+      kubectl patch pvc ${pvc} -n $namespace  --type=json -p='[{"op": "replace", "path": "/metadata/annotations/volume.beta.kubernetes.io~1storage-provisioner", "value":"'$new_driver_name'"}]'
+      kubectl patch pvc ${pvc} -n $namespace  --type=json -p='[{"op": "replace", "path": "/metadata/annotations/volume.kubernetes.io~1storage-provisioner", "value":"'$new_driver_name'"}]'
+      echo "Recreating pv: $pv"
+      kubectl create -f pv-${pv}.yaml
+      echo "Deleting annotation from pvc $pvc to trigger pvc/pv binding"
+      kubectl patch pvc ${pvc} -n $namespace --type=json -p='[{"op": "remove", "path": "/metadata/annotations/pv.kubernetes.io~1bind-completed"}]'
+      # kubectl patch pvc ${pvc} -n $namespace --type=json -p='[{"op": "remove", "path": "/metadata/annotations/pv.kubernetes.io~1bound-by-controller"}]'
```
Comment (apolovov) [inline hooks/migrate_csi_endpoint.sh:204]: ?

**Review by apolovov — CHANGES_REQUESTED** [review]

File: hooks/migrate_csi_endpoint.sh (hunk 6 above)
Comment (AleksZimin) [reply hooks/migrate_csi_endpoint.sh]: Move export AFFECTED_PVS_HASH="" to a position before the functions where it is set.

File: hooks/migrate_csi_endpoint.sh (hunk 7 above)
Comment (AleksZimin) [reply hooks/migrate_csi_endpoint.sh]: Fixed

File: hooks/migrate_csi_endpoint.sh (hunk 9 above)
Comment (AleksZimin) [reply hooks/migrate_csi_endpoint.sh]: Fixed

**Review by duckhawk — APPROVED** [review]

**Pull request merged**